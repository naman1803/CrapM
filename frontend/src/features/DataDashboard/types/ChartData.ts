// used for SeverityBarChart
export interface SeverityData {
    facilityOwner: string;
    severityLevel: string;
    severityScore: number;
}

// used for SeverityOverTimeChart
export interface SeverityOverTimeData extends SeverityData {
    date: string;
}

// one instance of ranking
export interface FacilityExceedanceData {
    facilityOwner: string;
    noOfExceedances: number;
}
