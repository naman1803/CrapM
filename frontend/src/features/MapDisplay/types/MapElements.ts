import { LatLngExpression } from 'leaflet';

export enum MapElementType {
  Marker = 'marker',
  Circle = 'circle',
}

// Represent a point on the map
interface MapElementBase {
  id: number;
  dataType: string;
  isSelectable: boolean;
}

interface MarkerElement extends MapElementBase {
  elementType: MapElementType.Marker;
  dataPoint: GeoPoint;
}

interface CircleElement extends MapElementBase {
  elementType: MapElementType.Circle;
  dataPoint: GeoCircle;
}

export interface GeoPoint {
  coordinates: LatLngExpression;
}

interface GeoCircle extends GeoPoint {
  radius: number;
  colour: string;
}

export interface LakeGeoPoint extends GeoPoint {
  id: number;
  name: string;
  additional_notes?: string;
}

export interface SewageGeoPoint extends GeoCircle {
  id: number;
  facility_owner: string;
}

export type MapElement = MarkerElement | CircleElement;
