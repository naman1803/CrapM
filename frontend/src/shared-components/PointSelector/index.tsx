export { default as PointSelector } from './PointSelector';