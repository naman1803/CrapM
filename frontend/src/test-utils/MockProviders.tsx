import { ThemeProvider } from '@mui/material';
import { MemoryRouter } from 'react-router-dom';
import { themeOptions } from '@app/theme';
import React from 'react';

// Wraps all providers for the app for testing
export const TestProviders = ({
  children,
  initialRoute = '/',
}: {
  children?: React.ReactNode;
  initialRoute?: string;
}) => {
  return (
    <ThemeProvider theme={themeOptions}>
      <MemoryRouter initialEntries={[initialRoute]}>{children}</MemoryRouter>
    </ThemeProvider>
  );
};
