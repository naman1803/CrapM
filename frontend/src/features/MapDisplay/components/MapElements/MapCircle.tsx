import { Circle } from 'react-leaflet';
import { MapElement, SewageGeoPoint } from '../../types/MapElements';
import React, { useState, useEffect, useRef } from 'react';
import { LatLngTuple } from 'leaflet';

interface MapElementCircleProps {
  element: MapElement;
  onElementSelect: (elementSelected: MapElement) => void;
  hoveredPoint: MapElement | null;
  selectedPoint: MapElement | null;
}
const MapElementCircle: React.FC<MapElementCircleProps> = ({
  element,
  onElementSelect,
  hoveredPoint,
  selectedPoint,
}) => {
  const datapoint = element.dataPoint as SewageGeoPoint;
  const [color, setColor] = useState('gray'); // Default color
  const baseColor = useRef('gray');

  // Change color based on hover
  useEffect(() => {
    if (
      selectedPoint &&
      selectedPoint.dataPoint.coordinates === datapoint.coordinates
    ) {
      setColor('red'); // Selected color
      baseColor.current = 'red';
    } else if (
      hoveredPoint &&
      hoveredPoint.dataPoint.coordinates === datapoint.coordinates
    ) {
      setColor('blue'); // Hover color
      baseColor.current = 'blue';
    } else {
      setColor(datapoint.colour); // Default color
      baseColor.current = datapoint.colour;
    }
  }, [selectedPoint, hoveredPoint, datapoint.coordinates, datapoint.colour]);

  return (
    <Circle
      radius={isNaN(datapoint?.radius) ? 20 : datapoint.radius * 20}
      color={datapoint.colour}
      center={datapoint.coordinates}
      data-testid={`Circle-${datapoint.coordinates as LatLngTuple}`}
      pathOptions={{ color }}
      eventHandlers={{
        ...(element.isSelectable
          ? {
              click: () => onElementSelect(element),
              mouseover: () => setColor('blue'), // Hover color
              mouseout: () => setColor(baseColor.current), // Revert to base color
            }
          : {}),
      }}
    ></Circle>
  );
};

export default MapElementCircle;
