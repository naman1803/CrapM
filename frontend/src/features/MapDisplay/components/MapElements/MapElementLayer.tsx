import { MapElementType, MapElement } from '../../types/MapElements';
import MapElementMarker from './MapMarker';
import MapElementCircle from './MapCircle';

export interface MapElementLayerProps {
  elements: MapElement[];
  onElementClick: (element: MapElement) => void;
  hoveredPoint: MapElement | null;
  selectedPoint: MapElement | null;
  zoom: number;
}

const MapElementLayer: React.FC<MapElementLayerProps> = ({
  elements,
  onElementClick,
  hoveredPoint,
  selectedPoint,
  zoom,
}) => {
  return (
    <>
      {elements.map((element) => {
        switch (element.elementType) {
          case MapElementType.Marker:
            return (
              <MapElementMarker
                key={`${element.id}-${element.elementType}`}
                element={element}
                onElementSelect={onElementClick}
                hoveredPoint={hoveredPoint}
                selectedPoint={selectedPoint}
                zoom={zoom}
              />
            );
          case MapElementType.Circle:
            return (
              <MapElementCircle
                key={`${element.id}-${element.elementType}`}
                element={element}
                onElementSelect={onElementClick}
                hoveredPoint={hoveredPoint}
                selectedPoint={selectedPoint}
              />
            );
          default:
            return <></>;
        }
      })}
    </>
  );
};
export default MapElementLayer;
