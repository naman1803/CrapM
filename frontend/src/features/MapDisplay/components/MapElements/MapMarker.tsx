import { Marker } from 'react-leaflet';
import { MapElement } from '../../types/MapElements';
import React, { useEffect, useState, useRef } from 'react';
import L from 'leaflet';

interface MapElementMarkerProps {
  element: MapElement;
  onElementSelect?: (elementSelected: MapElement) => void;
  hoveredPoint: MapElement | null;
  selectedPoint: MapElement | null;
  zoom: number;
}

const MapElementMarker: React.FC<MapElementMarkerProps> = ({
  element,
  onElementSelect,
  hoveredPoint,
  selectedPoint,
  zoom
}) => {
  const [color, setColor] = useState('gray');  // Default color
  const baseColor = useRef('gray');

  // Change color based on hover
  useEffect(() => {
    if (selectedPoint && selectedPoint.dataPoint.coordinates === element.dataPoint.coordinates) {
      setColor('red');  // Selected color
      baseColor.current = 'red';
    } else if (hoveredPoint && hoveredPoint.dataPoint.coordinates === element.dataPoint.coordinates) {
      setColor('blue');  // Hover color
      baseColor.current = 'blue';
    } else {
      setColor('gray');  // Default color
      baseColor.current = 'gray';
    }
  }, [selectedPoint, hoveredPoint, element.dataPoint.coordinates]);
  
  const MIN_SIZE = 7; 

  const markerSize = Math.max(MIN_SIZE, zoom*1.5);
  // Custom HTML-based icon with dynamic color
  const markerHtmlStyles = `
    background-color: ${color};
    width: ${markerSize}px;
    height: ${markerSize}px;
    display: block;
    left: -${markerSize / 2}px;  /* Center horizontally */
    top: -${markerSize / 2}px;   /* Center vertically */
    position: relative;
    border-radius: 3rem 3rem 0;
    transform: rotate(45deg);
    border: 1px solid #FFFFFF;
  `;

  const icon = L.divIcon({
    className: 'lake-marker',
    iconAnchor: [0, 10],
    popupAnchor: [0, -36],
    html: `<span style="${markerHtmlStyles}" />`,
  });

  return (
    <Marker
      position={element.dataPoint.coordinates}
      icon={icon}
      eventHandlers={{
        ...(element.isSelectable
          ? {
              click: () => onElementSelect && onElementSelect(element),
              mouseover: () => setColor('blue'), // Hover color
              mouseout: () => setColor(baseColor.current) // Revert to base color
            }
          : {})
      }}
    />

  );
};

export default MapElementMarker;