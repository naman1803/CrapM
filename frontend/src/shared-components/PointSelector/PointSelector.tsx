import { Box, Button, ButtonGroup, Grid2 } from '@mui/material';
import React from 'react';
import { SewageGeoPoint, LakeGeoPoint, MapElement } from '@/features/MapDisplay/types/MapElements';

export interface PointSelectorProps {
  onListElementClick: (element: MapElement) => void;
  points: MapElement[];
  type: 'lake' | 'sewage';
  onHoverPoint: (element: MapElement | null) => void;
}

const PointSelector: React.FC<PointSelectorProps> = ({ points, onListElementClick, type, onHoverPoint }) => {
  return (
    <Grid2 size={{ xs: 12, md: 6 }}>
      <Box
        sx={{
          display: 'flex',
          width: '100%',
          backgroundColor: 'rgba(0, 0, 0, 0.75)',
          border: '1px solid white',
          borderRadius: '10px'
        }}
        minHeight={"300px"}
        maxHeight={"300px"}
      >
        <ButtonGroup
          color="primary"
          orientation="vertical"
          sx={{
            flexGrow: 1,
            maxWidth: '100%',
            display: 'flex',
            flexDirection: 'column',
            overflowX: 'hidden',
          }}
        >
          {points.map((point) => {
            const isLake = type === 'lake';
            const dataPoint = point.dataPoint as (typeof isLake extends true ? LakeGeoPoint : SewageGeoPoint);

            return (
              <Button
                key={point.id}
                onClick={() => onListElementClick(point)}
                sx={{
                  textTransform: 'none',
                  border: 'none',
                  borderRadius: '0px',
                  color: 'white',
                  flexGrow: 1,
                  backgroundImage: `linear-gradient(to bottom, rgba(0, 0, 0, 0.75) 51%, white 49%)`,
                  backgroundPosition: 'top',
                  backgroundSize: '100% 200%',
                  transition: 'background-position 0.5s ease',
                  '&:hover': {
                    color: 'rgba(0, 0, 0, 0.75)',
                    backgroundPosition: 'bottom',
                  },
                }}
                onMouseEnter={() => onHoverPoint(point)}
                onMouseLeave={() => onHoverPoint(null)}
              >
                {isLake ? (point.dataPoint as LakeGeoPoint).name : (dataPoint as SewageGeoPoint).facility_owner}
              </Button>
            );
          })}
        </ButtonGroup>
      </Box>
    </Grid2>
  );
};

export default PointSelector;