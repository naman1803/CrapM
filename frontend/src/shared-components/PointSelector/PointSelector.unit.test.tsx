import { render } from '@testing-library/react';
import { TestProviders } from '@/test-utils/MockProviders';
import PointSelector, { PointSelectorProps } from './PointSelector';
import { LatLngExpression } from 'leaflet';
import { LakeGeoPoint, MapElementType } from '@/features/MapDisplay/types/MapElements';

const mockProps: PointSelectorProps = {
  type: 'lake',
  points: [
    {
      elementType: MapElementType.Marker,
      dataType: 'marker',
      id: 1,
      dataPoint: {
        id: 1,
        name: "Mynor Lake",
        coordinates: [0.0, 0.0] as LatLngExpression,
        additional_notes: undefined
      } as LakeGeoPoint,
      isSelectable: true,
    },{
      elementType: MapElementType.Marker,
      dataType: 'marker',
      id: 2,
      dataPoint: {
        id: 2,
        name: "Red Lake",
        coordinates: [3.0, 3.0] as LatLngExpression,
        additional_notes: undefined
      } as LakeGeoPoint,
      isSelectable: true,
    }
  ],
  onListElementClick: jest.fn(),
  onHoverPoint: jest.fn(),
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('PointSelector', () => {
  it('renders correctly', () => {
    const { asFragment } = renderPointSelector(mockProps);
    expect(asFragment()).toMatchSnapshot();
  });
  
  it('renders elements', () => {
    const { getByRole } = renderPointSelector(mockProps);
    expect(getByRole('button', { name: 'Mynor Lake' })).toBeInTheDocument();
    expect(getByRole('button', { name: 'Red Lake' })).toBeInTheDocument();
  });
  
  it('calls onListElementClick when list button is clicked', () => {
    const { getAllByRole } = renderPointSelector(mockProps);
    getAllByRole('button')[0].click();
    expect(mockProps.onListElementClick).toHaveBeenCalled();
    getAllByRole('button')[1].click();
    expect(mockProps.onListElementClick).toHaveBeenCalled();
  });
});

const renderPointSelector = (props: PointSelectorProps) => {
  return render(
    <TestProviders>
      <PointSelector {...props} />
    </TestProviders>,
  );
};