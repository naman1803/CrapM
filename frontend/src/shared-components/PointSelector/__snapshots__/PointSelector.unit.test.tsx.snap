// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointSelector renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-6 css-1odx5a8-MuiGrid2-root"
  >
    <div
      class="MuiBox-root css-1eruiwu"
    >
      <div
        class="MuiButtonGroup-root MuiButtonGroup-outlined MuiButtonGroup-vertical MuiButtonGroup-colorPrimary css-fbq7xj-MuiButtonGroup-root"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedVertical MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedVertical MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedVertical MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedVertical MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-firstButton css-jx6btc-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Mynor Lake
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedVertical MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedVertical MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButtonGroup-grouped MuiButtonGroup-groupedVertical MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedVertical MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-lastButton css-jx6btc-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Red Lake
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
